added gene_name array to the zarr files and also updated the specs in…

kishorikonwar opened the pull request:
added gene_name array to the zarr files and also updated the specs in the zarr specifications

### Purpose
Code changes to add gene_name  array to the zarr files for Optimus, such that the ith element in the gene_id corresponds to the ith element in the gene_name array.  An example,  a zarr file with gene_name would look like 

```
>>> import zarr
>>> a = zarr.open("test2.zarr")
>>> a.tree()
/
 └── expression_matrix
     ├── cell_id (397718,) <U40
     ├── cell_metadata_numeric (397718, 32) float32
     ├── cell_metadata_numeric_name (32,) <U40
     ├── expression (397718, 58347) float32
     ├── gene_id (58347,) <U40
     ├── gene_metadata_numeric (58347, 26) float32
     ├── gene_metadata_numeric_name (26,) <U40
     └── gene_name (58347,) <U40
>>> a['expression_matrix']['gene_id'][0:25]
array(['ENSG00000223972.5', 'ENSG00000227232.5', 'ENSG00000278267.1',
       'ENSG00000243485.5', 'ENSG00000284332.1', 'ENSG00000237613.2',
       'ENSG00000268020.3', 'ENSG00000240361.2', 'ENSG00000186092.5',
       'ENSG00000238009.6', 'ENSG00000239945.1', 'ENSG00000233750.3',
       'ENSG00000268903.1', 'ENSG00000269981.1', 'ENSG00000239906.1',
       'ENSG00000241860.6', 'ENSG00000222623.1', 'ENSG00000241599.1',
       'ENSG00000279928.2', 'ENSG00000279457.4', 'ENSG00000273874.1',
       'ENSG00000228463.10', 'ENSG00000236679.2', 'ENSG00000236601.2',
       'ENSG00000237094.12'], 
      dtype='<U40')
>>> a['expression_matrix']['gene_name'][0:25]
array(['DDX11L1', 'WASH7P', 'MIR6859-1', 'MIR1302-2HG', 'MIR1302-2',
       'FAM138A', 'AL627309.6', 'OR4G11P', 'OR4F5', 'AL627309.1',
       'AL627309.3', 'CICP27', 'AL627309.7', 'AL627309.8', 'AL627309.2',
       'AL627309.5', 'RNU6-1100P', 'AL627309.4', 'FO538757.2',
       'FO538757.1', 'MIR6859-2', 'AP006222.1', 'RPL23AP24', 'AL732372.1',
       'AL732372.2'], 
      dtype='<U40')
>>> 
```

Note that gene_id ENSG00000223972.5' corresponds to gene_name WASH7P, and so on.





- No issue is linked to this PR.

### Changes
<!-- Please list out what major changes were made in this PR to address the issue: -->

- No changes.

### Review Instructions
<!-- Please provide instructions about how should a reviewer test/verify the changes in this PR: -->

- No instructions.

kbergin commented:
@ambrosejcarr I believe Polina's changes updated the memory bug that we were running into, and Kishori's updates the zarr outputs to include gene name as a metadata in addition to gene id (added previously)

mckinsel commented on docker/python3-scientific/create_zarr_optimus.py:
```diff
@@ -224,6 +225,40 @@ def add_cell_metrics(data_group, input_path, cell_ids, verbose=False):
+        else:
+            fpin = open(gtf_file, 'r')
+    except FileNotFoundError:
+        print("GTF File not found")
```
Do you really want to swallow the exception here? Seems like the function should fail at this point.

kishorikonwar replied:
I agree! removed the exception-handling part.

mckinsel commented on docker/python3-scientific/create_zarr_optimus.py:
```diff
@@ -224,6 +225,40 @@ def add_cell_metrics(data_group, input_path, cell_ids, verbose=False):
+        gene_name_res = re.search(r'gene_name ([^;]*);', line)
+
+        if gene_id_res and gene_name_res:
+            gene_id = re.sub("\"", "", gene_id_res.group(1))
```
Would be better to use single quotes rather than escaping: `'"'`. Also consider just using `replace` rather than `re.sub`; it's a little simpler.

kishorikonwar replied:
Agree! Fixed!

mckinsel commented:
Could `gene_name` just be an instance of gene string metadata?

rexwangcc commented on docker/python3-scientific/create_zarr_optimus.py:
```diff
@@ -224,6 +225,40 @@ def add_cell_metrics(data_group, input_path, cell_ids, verbose=False):
+    """ Creates a map from gene_id to gene_name by reading in the GTF file
+
+    Args:
+        annotation_file (string): annotation file 
```
```suggestion
        annotation_file (str): annotation file 
```

rexwangcc commented on docker/python3-scientific/create_zarr_optimus.py:
```diff
@@ -224,6 +225,40 @@ def add_cell_metrics(data_group, input_path, cell_ids, verbose=False):
+        annotation_file (string): annotation file 
+
+    Return:
+        gene_id_name_map (dict[string]->string): dictonary gene ids to gene names
```
```suggestion
        gene_id_name_map (Dict[str, str]): dictonary gene ids to gene names
```

rexwangcc commented on docker/python3-scientific/create_zarr_optimus.py:
```diff
@@ -224,6 +225,40 @@ def add_cell_metrics(data_group, input_path, cell_ids, verbose=False):
+        if gtf_file.endswith(".gz"):
+            fpin = gzip.open(gtf_file, 'rt')
+        else:
+            fpin = open(gtf_file, 'r')
```
I'm thinking if we use something like:
```python
with gzip.open(gtf_file, 'rt') if gtf_file.endswith(".gz") else open(gtf_file, 'r') as fpin:
    blah
```
we could get rid of the `fpin.close()` and make it "safer". Please test if it works if you decide to go this way since I didn't test it!

kishorikonwar replied:
Thanks! It worked!

rexwangcc commented on docker/python3-scientific/create_zarr_optimus.py:
```diff
@@ -224,6 +225,40 @@ def add_cell_metrics(data_group, input_path, cell_ids, verbose=False):
+            gene_id = re.sub("\"", "", gene_id_res.group(1))
+            gene_name = re.sub("\"", "", gene_name_res.group(1))
+            gene_id_name_map[gene_id] = gene_name
+    fpin.close()
```
please see my comment above

kishorikonwar replied:
Changed!

rexwangcc commented on docker/python3-scientific/create_zarr_optimus.py:
```diff
@@ -266,6 +301,19 @@ def add_expression_counts(data_group, args):
+    # add the gene names for the gene ids 
+    if args.annotation_file:
+        gene_id_name_map = create_gene_id_name_map(args.annotation_file)
+        gene_names = [gene_id_name_map[gene_id] if gene_id in gene_id_name_map else "" for gene_id in gene_ids]
```
Could we simplify this with:
```suggestion
        gene_names = [gene_id_name_map.get(gene_id, "") for gene_id in gene_ids]
```
?

rexwangcc commented on pipelines/optimus/Optimus.wdl:
```diff
@@ -213,6 +213,7 @@ workflow Optimus {
     call ZarrUtils.OptimusZarrConversion{
       input:
         sample_id=sample_id,
+        annotation_file=annotations_gtf,
```
```suggestion
        annotation_file = annotations_gtf,
```

rexwangcc commented on library/tasks/ZarrUtils.wdl:
```diff
@@ -93,6 +96,7 @@ task OptimusZarrConversion {
     set -euo pipefail
 
     python3 /tools/create_zarr_optimus.py \
+       --annotation_file  ${annotation_file}\
```
```suggestion
       --annotation_file ${annotation_file}\
```

kishorikonwar commented:
Addressed the suggestions from the reviewers

>>> import zarr
>>> a=zarr.open("test2.zarr")
>>> a.tree()
/
 └── expression_matrix
     ├── cell_id (308068,) <U40
     ├── cell_metadata_numeric (308068, 32) float32
     ├── cell_metadata_numeric_name (32,) <U40
     ├── expression (308068, 58347) float32
     ├── gene_id (58347,) <U40
     ├── gene_metadata_numeric (58347, 26) float32
     ├── gene_metadata_numeric_name (26,) <U40
     ├── gene_metadata_string (1, 58347) <U40
     └── gene_metadata_string_name (1,) <U40

>>> a['expression_matrix']['gene_metadata_string_name'].shape
(1,)

>>> a['expression_matrix']['gene_metadata_string_name'][:]
array(['gene_name'], 
      dtype='<U40')

>>> a['expression_matrix']['gene_metadata_string'].shape
(1, 58347)

>>> a['expression_matrix']['gene_metadata_string'][:20]
array([['DDX11L1', 'WASH7P', 'MIR6859-1', ..., 'AC007325.2', 'U6', 'U1']], 
      dtype='<U40')

mckinsel commented on docs/matrix_format_spec.md:
```diff
@@ -45,6 +45,15 @@ UUID generated by the DCP or a user-submitted identifier.
+  - shape: `(n_genes,)`
+  - compressor: `Blosc(cname='lz4', clevel=5, shuffle=SHUFFLE, blocksize=0)`
+  - chunk shape: `(10000,)`
+  - dtype: `U40`
```
I think you can remove this part from the spec.

mckinsel commented:
Are the changes to disk space in the two wdls still needed after #228?
